gateway/api/views/files.py

```diff
@@ -276,48 +288,152 @@ def provider_list(self, request):
         return Response({"results": files})
 
     @action(methods=["GET"], detail=False)
-    def download(self, request):  # pylint: disable=invalid-name
-        """Download selected file."""
-        # default response for file not found, overwritten if file is found
-        response = Response(
-            {"message": "Requested file was not found."},
-            status=status.HTTP_404_NOT_FOUND,
-        )
+    def download(self, request):
+        """
+        It returns a file from user paths:
+            - username/
+            - username/provider_name/function_title
+        """
         tracer = trace.get_tracer("gateway.tracer")
         ctx = TraceContextTextMapPropagator().extract(carrier=request.headers)
         with tracer.start_as_current_span("gateway.files.download", context=ctx):
-            requested_file_name = request.query_params.get("file")
-            provider_name = request.query_params.get("provider")
-            if requested_file_name is not None:
-                user_dir = request.user.username
-                if provider_name is not None:
-                    if self.check_user_has_provider(request.user, provider_name):
-                        user_dir = provider_name
-                    else:
-                        return response
-                # look for file in user's folder
-                filename = os.path.basename(requested_file_name)
-                user_dir = os.path.join(
-                    sanitize_file_path(settings.MEDIA_ROOT),
-                    sanitize_file_path(user_dir),
+            username = request.user.username
+            requested_file_name = sanitize_file_name(
+                request.query_params.get("file", None)
+            )
+            provider_name = sanitize_name(request.query_params.get("provider", None))
```

Comment:
Should we check if the user has provider access if provider is provided?

Reply:
Yes, it's being checked inside `self.get_function`. If the user doesn't have access `get_function` returns a `None` and the view returns a `404`. 

We log in case the user doesn't have access in the traces internally.